### PR TITLE
Hopcroft Tarjan planarity test algorithm

### PR DESCRIPTION
**Pull request by KarolinaRezkova**
implementation of Hopcroft Tarjan algorithm for testing if graph is planar

### CODE REVIEW

**Comment by KarolinaRezkova** [comment]
Thank you for your comments, I went through them and tried to fix all the issues you pointed out.

I haven't managed to work out the certificate for the result yet.

Since the algorithm is not based on finding  a Kuratowski subgraph, I don't think it is possible to somehow derive it from the result. As far as I know, it should be possible to modify the algorithm to return a planar embedding, but I haven't focused on that part yet.

**Comment by jkinable** [comment]
Thanks for the changes. The copyright header is not yet correct. Please copy and paste the **full** header from any of the source files already present in jgrapht. Only modify your name and the year 2017-2017.

**Comment by jkinable** [comment]
As it turns out, the algorithmic library LEDA has an implementation of the Hopcroft Tarjan planarity testing algorithm. Their implementation is based on this [paper](https://pdfs.semanticscholar.org/0cf9/43fa4b2593e21d35ccaa31e0a9ec6dfde2fd.pdf). More details about their implementation can be found on chapter 8 of the [LEDA book](https://people.mpi-inf.mpg.de/~mehlhorn/ftp/LEDAbook/EmbeddedGraphs.ps). This is definitely worth looking into, as their implementation are often very efficient. As expected, they indeed produce a Kuratowski subgraph to certify that a planar embedding does not exist.

Also, in the LEDA manual they state the following fact:

> Interesting Fact 2: A planar graph G=(V,E) (without parallel edges and selfloops) can have at most 3|V|-6 edges.
You may already have this check in your code. If not, then this is something easy (and cheap) to check.

File: jgrapht-core/src/main/java/org/jgrapht/alg/planarity/HopcroftTarjanPlanarityInspector.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+     *
+     * @return true if graph is planar, false otherwise
+     */
+    public boolean isPlanar() {
+        
+        if (this.graph.vertexSet().size()<=4){
+            return true;
+        }
+
+        if ((this.graph.edgeSet().size()) > (this.graph.vertexSet().size() * 3 - 6)) {
```
Comment (KarolinaRezkova) [inline jgrapht-core/src/main/java/org/jgrapht/alg/planarity/HopcroftTarjanPlanarityInspector.java]: the edge size check is located here

**Comment by KarolinaRezkova** [comment]
Thank you, I hope the header is ok now. 

I will have a look at the LEDA implementation you suggested and see if I am able to modify my code to return the certificate of result.

**Comment by jkinable** [comment]
Unfortunately, the header is still not good...
Compare your header:
`/* (C) Copyright 2017-2017, by Karolina Rezkova and Contributors`.
with for instance the header in GraphTests
```
/*
 * (C) Copyright 2003-2017, by Barak Naveh, Dimitrios Michail and Contributors.

```
Notice the **"/*"** on the first line....
I ran a search on the jgrapht directory and I couldn't find any file which started with "/* (C)", so I am not sure where you copy/pasted that one from? Please don't modify the header beyond the year and the author's name.

**Comment by jkinable** [comment]
Any updates on this pull request? We are getting close to the next release....

**Comment by jkinable** [comment]
Any updates on this PR? It would be pity to abandon it :(

**Comment by rkrishnasanka** [comment]
@jkinable What are the changes remaining on this that need to be done ?

**Comment by jkinable** [comment]
@rkrishnasanka At the very least, the merge conflicts must be resolved. Also the header format needs to be fixed. Finally @KarolinaRezkova indicated that she would look into the possibility to generate certificates, but we unfortunately never heard back.

**Comment by rkrishnasanka** [comment]
I could definitely fix the headers and take a stab at the merge conflicts in 2 weeks. I would be useful to have the planarity check in jGrapht for some of my work.

**Comment by waahm7** [comment]
I would like to pick this up.

**Comment by jsichi** [comment]
I'm gonna close this one since we now have #795 and there's probably no point adding Hopcroft-Tarjan once we have Boyer-Myrvold.

**Pull request closed without merging**